### ISSUE TITLE
Support for abstract / polymorphic classes

### ISSUE THREAD

**Issue by williamboxhall** [avro-kotlin/avro4k#108]
According to the [kotlinx serialization docs](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md#registered-subclasses) it should be possible to register the subtypes of an abstract class as part of the `serializerModule`. However, it doesn't seem to work:

```kotlin
@file:UseContextualSerialization
...
    val module = SerializersModule {
        polymorphic(SkillsCoachEvent::class) {
            subclass(AccountSettingsEvent::class)
        }
    }
    val schema = Avro(configuration = AvroConfiguration(SnakeCaseNamingStrategy), serializersModule = module).schema(Event.serializer())

...

@Serializable
data class Event(
    val eventId: UUID,
    val createdAt: DateTime,
    val aggregateId: UUID,
    val sequence: Int,
    val aggregateType: String,
    val eventType: String,
    val metadata: ManagerLabMetadata,
    @Polymorphic @Contextual val body: SkillsCoachEvent
)

...

@Serializable
@Polymorphic
abstract class SkillsCoachEvent : DomainEvent
```
As you can see, I also naively tried adding both `@Polymorphic` and `@Contextual` and `@file:UseContextualSerialization` clutching at straws that they'D offer the hints necessary for Avro4K to treat this as `SerialKind.CONTEXTUAL` instead of `SerialKind.OPEN`, to no available. I suspect that Avro4K might need to introduce a new `SerialKind.POLYMORPHIC` or something like that, to match what's in kotlinx serialization.

```
Exception in thread "main" kotlinx.serialization.SerializationException: Unsupported type kotlinx.serialization.Polymorphic<SkillsCoachEvent> of OPEN
	at com.github.avrokotlin.avro4k.schema.SchemaForKt.schemaFor(SchemaFor.kt:206)
	at com.github.avrokotlin.avro4k.schema.SchemaForKt.schemaFor(SchemaFor.kt:187)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.buildField(ClassSchemaFor.kt:78)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.dataClassSchema(ClassSchemaFor.kt:63)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.schema(ClassSchemaFor.kt:43)
	at com.github.avrokotlin.avro4k.Avro.schema(Avro.kt:264)
	at com.github.avrokotlin.avro4k.Avro.schema(Avro.kt:269)
	at events.AvroSchemaGeneratorKt.main(AvroSchemaGenerator.kt:37)
```

**Comment by williamboxhall** [comment]
Thanks @thake that's a fair call. I'm not sure if I'll be able to absorb getting up to speed in avro4ks internals and then doing this PR while dealing with our internal deadlines, but I will definitely try :).

In the meantime, our workaround is probably to just use properly sealed classes, which causes some other significant compromises but gives us a path forward.

**Comment by williamboxhall** [comment]
I've taken a stab at making a PR for this: https://github.com/avro-kotlin/avro4k/pull/109

**Comment by thake** [comment]
Implemented with #110
